app/controllers/application_controller.rb

```diff
@@ -1,14 +1,38 @@
 class ApplicationController < ActionController::Base
   protect_from_forgery with: :exception
 
-  before_action :find_user
+  before_action :find_user, :require_login
 
   def render_404
     # DPR: supposedly this will actually render a 404 page in production
     raise ActionController::RoutingError.new('Not Found')
   end
 
-private
+  protected
+  def require_login
+    @user = User.find_by(id: session[:user_id])
+    unless @user
+      flash[:status] = :failure
+      flash[:result_text] = "You must be logged in to do that!"
+      redirect_to root_path
+    end
+  end
+
+  def save_and_flash(model)
+
+
+    if model
+      flash[:status] = :success
+      flash[:message] = "Successfully saved #{model.class} #{Model.id}"
+    else
```

Comment:
This failed for me the first time I logged in - it looks like you're reporting `Model.id`, when you probably want `model.id`. This is exactly the sort of problem that thorough testing should be able to catch.